GenericList/list.h

```diff
@@ -0,0 +1,263 @@
+#pragma once
+
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <memory>
+#include <cassert>
+#include <functional>
+#include <algorithm>
+#include <type_traits>
+
+// Never use header-wide using directives ("using namespace") in the header!!
+// Explanation: https://stackoverflow.com/questions/5849457/using-namespace-in-c-headers
+// Basically, when this header is included, the "using namespace" will be included too, possibly causing un-intended clashes in variable/function/etc.. namess
+// Note: as long as code doesn't #include a .cpp file, it's fine to use using directives in that source file
+
+//Generic type, allow for stateful Allocator if user desires it
+template <typename T, typename Allocator = std::allocator<T>>
+class List {
+    static_assert(!std::is_void_v<T>, "void type is not allowed");
+    static_assert(!std::is_reference_v<T>, "reference type is not allowed");
+    //C++ Standard �8.3.2/4: There shall be no references to references, no arrays of references, and no pointers to references.
+
+
+    //note: elements are value copies of original objects (copy-by-value, not copy-by-reference)
+public:
+    List() : dataAllocator(), data(nullptr), capacity(0), count(0) {
+
+    }
+
+    List(Allocator const& alloc) : dataAllocator(alloc), data(nullptr), capacity(0), count(0) {
+    
+    }
+
+    friend void swap(List& first, List& second) noexcept {
+        //Purely swap data: since we want the states to remain the same as beforehand, just switched, explicitely use std::swap in case swap(Allocator&, Allocator&) somehow moves memory around
+        //Note: since data is a pointer value, it cannot have a "custom swap friend function" - aka, let's just use std::swap everywhere
+        std::swap(first.dataAllocator, second.dataAllocator);
+        std::swap(first.capacity, second.capacity);
+        std::swap(first.count, second.count);
+        std::swap(first.data, second.data);
+    }
+
+    //Rule of 3
+    ~List() { //Destructor
+        Clear();
+        dataAllocator.deallocate(data, capacity);
+    }
+
+    //Copy Constructor - dataAllocator is not copied over as the data allocated is not the same (different memory address, deep copy)
+    //TODO: look into std::allocator_traits
+    List(const List& other) : dataAllocator(other.dataAllocator), data(CreateDeepCopy(other, dataAllocator)),
+                              capacity(other.capacity), count(other.count) {}
+
+    List& operator=(const List& other) { //Copy assignement - keep dataAllocator the same as it was
+        if (this != &other) {
+            Clear();
+            Capacity(other.capacity);
```

Comment:
Sorry, I left this comment unposted on my work account. You should deallocate the data and copy the allocator across as well.
```c++
Clear();
dataAllocator.deallocate(data, capacity);
dataAloocator = other.dataAllocator;
Capacity(other.capacity);
```

Reply:
Can't find anything in cpp docs about `deallocate(.., 0)` expected behaviour... (in some cases, `deallocate(..,0) is called without `allocate` ever being called).
Since everything works, I'm assuming that this isn't an issue, but ideally would add an `isAllocated` boolean to ensure `deallocate` is only called when data is already allocated.

Reply:
You can assume there's no allocation when `capacity == 0 ` so you can do:
```c++
if (capacity > 0) {
    dataAllocator.deallocate(data, capacity);
}
```
I think it's also typical that deallocators will deal with `nullptr`, but I don't really have any proof to back that up with 😅

Reply:
Ok nevermind, did a quick search and it turns out the standard does imply that you can't pass `nullptr` to deallocate. But typically `capacity > 0` implies not null so you can do
```c++
if (capacity > 0) {
    assert(data != nullptr);
    dataAllocator.deallocate(data, capacity);
}
```